ambry-store/src/main/java/com.github.ambry.store/DiskManager.java

```diff
@@ -311,11 +312,23 @@ boolean addBlobStore(ReplicaId replica) {
       if (!running) {
         logger.error("Failed to add {} because disk manager is not running", replica.getPartitionId());
       } else {
+        // Clean up existing dir associated with this replica to add. Here we re-create a new store because we don't
+        // know the state of files in old directory. (The old directory was created last time when adding this replica
```

Comment:
Just validating my though?
Is the assumption here that, if there is some old state of a replica on a node's storage, and storage manager or disk manager don't know about it, then it could be a corrupted state, and hence remove it.

The existing state might help in case where let's say a node crashed and comes back up after sometime, and is assigned same replica back again. In this case, even though there is some state on disk, we will delete it. The state on disk might have helped this replica in catching up with existing replicas quicker. But I assuming that we are ignoring this scenario because we don't have a way to identify this scenario. Is this correct?

Reply:
You are right. `The state on disk might have helped this replica in catching up with existing replicas quicker` this is true when the added replica was previously in BOOTSTRAP -> STANDBY and was in the middle of caught-up. Note that, if this added replica was ready in BOOTSTRAP -> STANDBY before node crashed, that means it should have already updated InstanceConfig in Helix (this is the last step in OFFLINE -> BOOTSTRAP). So, after node reboots, storage manager/disk manager should be aware of this replica and don't attempt to add it again. 
The failure I want to address in this PR occurs before new added replica updates InstanceConfig, so the new replica may be created but still empty. We don't have any loss by deleting its dir when node is restarted. (we can discuss it offline if this didn't answer your question)